File: exam_in_class/3/digital_library.py
# author: jamie
# email: jinjiedeng.jjd@gmail.com
class Node:
    def __init__ (self, key):
        self.key = key
        self.left = None
        self.right = None

class BinarySearchTree:
    def __init__ (self):
        self.root = None
        self.size = 0

    def Empty (self):
        return (self.root is None)

    def Size (self):
        return self.size

    def Insert (self, key):
        if not self.Search(key):
            self.size += 1
            if self.root:
                node = self.root
                while node:
                    if key < node.key:
                        if node.left:
                            node = node.left
                        else:
                            node.left = Node(key)
                            break
                    else:
                        if node.right:
                            node = node.right
                        else:
                            node.right = Node(key)
                            break
            else:
                self.root = Node(key)

    def Delete (self, key):
        if self.Search(key):
            self.size -= 1
            parent = None
            node = self.root
            while node:
                if key < node.key:
                    parent = node
                    node = node.left
                elif key > node.key:
                    parent = node
                    node = node.right
                else:
                    if node.left:
                        self.DeletePredecessorInstead(node)
                    else:
                        if parent:
                            if node == parent.left:
                                parent.left = node.right
                            else:
                                parent.right = node.right
                        else:
                            self.root = self.root.right
                    break

    def DeletePredecessorInstead (self, node):
        parent_pred = node
        pred = node.left
        while pred.right:
            parent_pred = pred
            pred = pred.right
        node.key = pred.key
        if parent_pred == node:
            parent_pred.left = pred.left
        else:
            parent_pred.right = pred.left

    def Search (self, key):
        node = self.root
        while node:
            if key < node.key:   node = node.left
            elif key > node.key: node = node.right
            else:                return True
        return False

    def Min (self):
        node = self.root
        while node and node.left: node = node.left
        return 0 if node is None else node.key

    def Max (self):
        node = self.root
        while node and node.right: node = node.right
        return 0 if node is None else node.key

    def Successor (self, key):
        succ = None
        node = self.root
        while node:
            if key >= node.key:
                node = node.right
            else:
                succ = node
                node = node.left
        return 0 if succ is None else succ.key

if __name__ == "__main__":
    n, m = map(int, input().split())
    B = [0]
    B.extend(map(int, input().split()))
    q = int(input())
    type_user = [BinarySearchTree() for _ in range(n + 1)]
    user_type = [BinarySearchTree() for _ in range(m + 1)]
    for _ in range(q):
        uid, tid = map(int, input().split())
        if not user_type[uid].Search(tid):
            if B[tid]:
                B[tid] -= 1
                user_type[uid].Insert(tid)
                type_user[tid].Insert(uid)
            else:
                revoked_uid = type_user[tid].Successor(uid) if uid < type_user[tid].Max() else type_user[tid].Max()
                user_type[revoked_uid].Delete(tid)
                user_type[uid].Insert(tid)
                type_user[tid].Delete(revoked_uid)
                type_user[tid].Insert(uid)
        print(user_type[uid].Size())
    for types in user_type[1:]:
        if types.Empty():
            print(0)
        else:
            rental_status = []
            while not types.Empty():
                min = types.Min()
                rental_status.append(min)
                types.Delete(min)
            print(" ".join(map(str, rental_status)))

File: exam_in_class/3/digital_library.cc
/*
* author: jamie
* email: jinjiedeng.jjd@gmail.com
*/
#include <iostream>
#include <set>
#include <vector>

int main ()
{
  uint64_t n, m; std::cin >> n >> m;
  std::vector<uint64_t> B(n + 1);
  for (uint64_t tid {1}; tid != B.size(); ++tid) { std::cin >> B[tid]; }
  uint64_t q; std::cin >> q;
  std::vector<std::set<uint64_t>> type_user(n + 1), user_type(m + 1);
  while (q--)
  {
    uint64_t uid, tid; std::cin >> uid >> tid;
    if (user_type[uid].count(tid) == 0)
    {
      if (B[tid] == 0)
      {
        auto revoked_uid {(uid < *type_user[tid].rbegin()) ? *type_user[tid].upper_bound(uid) : *type_user[tid].rbegin()};
        user_type[revoked_uid].erase(tid);
        type_user[tid].erase(revoked_uid);
      } else { --B[tid]; }
      user_type[uid].insert(tid);
      type_user[tid].insert(uid);
    }
    std::cout << user_type[uid].size() << "\n";
  }
  for (uint64_t uid {1}; uid != user_type.size(); ++uid)
  {
    if (user_type[uid].size())
    {
      auto it {user_type[uid].begin()};
      std::cout << *it++;
      while (it != user_type[uid].end()) { std::cout << " " << *it++; }
    }
    else { std::cout << "0"; }
    std::cout << "\n";
  }
  return 0;
}

/*
* author: jamie
* email: jinjiedeng.jjd@gmail.com
*/
#include <iostream>
#include <memory>
#include <vector>

struct Node
{
  Node (uint64_t k = 0): key {k}, left_subtree_size {} {}

  uint64_t key;
  uint64_t left_subtree_size;
  std::unique_ptr<Node> left;
  std::unique_ptr<Node> right;
};

class BinarySearchTree
{
public:

  bool Empty () const;
  uint64_t Size () const;
  void Insert (uint64_t key);
  void Delete (uint64_t key);
  bool Search (uint64_t key) const;
  uint64_t Min () const;
  uint64_t Max () const;
  uint64_t Successor (uint64_t key) const;

private:

  void DeletePredecessorInstead (Node* node);

  std::unique_ptr<Node> root_;
  uint64_t size_;

};

bool BinarySearchTree::Empty () const
{
  return (root_ == nullptr);
}

uint64_t BinarySearchTree::Size () const
{
  return size_;
}

void BinarySearchTree::Insert (uint64_t key)
{
  if (!Search(key))
  {
    ++size_;
    if (!root_) { root_ = std::make_unique<Node>(key); return; }
    Node* node {root_.get()};
    while (node)
    {
      if (key < node->key)
      {
        ++(node->left_subtree_size);
        if (node->left) { node = node->left.get(); }
        else            { node->left = std::make_unique<Node>(key); return; }
      }
      else
      {
        if (node->right) { node = node->right.get(); }
        else             { node->right = std::make_unique<Node>(key); return; }
      }
    }
  }
  return;
}

void BinarySearchTree::DeletePredecessorInstead (Node* node)
{
  Node* parent_pred {node};
  Node* pred {node->left.get()};
  while (pred->right) { parent_pred = pred; pred = pred->right.get(); }
  node->key = pred->key;
  if (parent_pred == node) { node->left.reset(pred->left.release()); }
  else                     { parent_pred->right.reset(pred->left.release()); }
  return;
}

void BinarySearchTree::Delete (uint64_t key)
{
  if (Search(key))
  {
    --size_;
    Node* parent {};
    Node* node {root_.get()};
    while (node)
    {
      if (key < node->key)
      {
        --(node->left_subtree_size);
        parent = node; node = node->left.get();
      }
      else if (key > node->key)
      {
        parent = node; node = node->right.get();
      }
      else
      {
        if (node->left)
        {
          --(node->left_subtree_size);
          DeletePredecessorInstead(node);
        }
        else
        {
          if (parent)
          {
            if (node == parent->left.get()) { parent->left.reset(node->right.release()); }
            else                            { parent->right.reset(node->right.release()); }
          }
          else { root_.reset(root_->right.release()); }
        }
        break;
      }
    }
  }
  return;
}

bool BinarySearchTree::Search (uint64_t key) const
{
  Node* node {root_.get()};
  while (node)
  {
    if (key < node->key)      { node = node->left.get(); }
    else if (key > node->key) { node = node->right.get(); }
    else                      { return true; }
  }
  return false;
}

uint64_t BinarySearchTree::Min () const
{
  Node* node {root_.get()};
  while (node && node->left) { node = node->left.get(); }
  if (node) { return node->key; }
  return 0;
}

uint64_t BinarySearchTree::Max () const
{
  Node* node {root_.get()};
  while (node && node->right) { node = node->right.get(); }
  if (node) { return node->key; }
  return 0;
}

uint64_t BinarySearchTree::Successor (uint64_t key) const
{
  Node* succ {};
  Node* node {root_.get()};
  while (node)
  {
    if (key >= node->key) { node = node->right.get(); }
    else                  { succ = node; node = node->left.get(); }
  }
  if (succ) { return succ->key; }
  return 0;
}

int main ()
{
  uint64_t n, m; std::cin >> n >> m;
  std::vector<uint64_t> B(n + 1);
  for (uint64_t tid {1}; tid != B.size(); ++tid) { std::cin >> B[tid]; }
  uint64_t q; std::cin >> q;
  std::vector<BinarySearchTree> type_user(n + 1), user_type(m + 1);
  while (q--)
  {
    uint64_t uid, tid; std::cin >> uid >> tid;
    if (!user_type[uid].Search(tid))
    {
      if (B[tid] == 0)
      {
        auto revoked_uid {(uid < type_user[tid].Max()) ? type_user[tid].Successor(uid) : type_user[tid].Max()};
        user_type[revoked_uid].Delete(tid);
        type_user[tid].Delete(revoked_uid);
      } else { --B[tid]; }
      user_type[uid].Insert(tid);
      type_user[tid].Insert(uid);
    }
    std::cout << user_type[uid].Size() << "\n";
  }
  for (uint64_t uid {1}; uid != user_type.size(); ++uid)
  {
    if (!user_type[uid].Empty())
    {
      while (!user_type[uid].Empty())
      {
        auto min {user_type[uid].Min()}; user_type[uid].Delete(min);
        std::cout << min;
        std::cout << ((user_type[uid].Empty()) ? "\n" : " ");
      }
    }
    else { std::cout << "0\n"; }
  }
  return 0;
}